File: c++/327. Count of Range Sum.cpp
class BIT {
public:
    BIT(int _n = 0):  n(_n) {
        c.resize(n + 1, 0);
    }
    
    void add(int idx, int val) {
        if (idx < 1 || idx > n) {
            return;
        }
        while (idx <= n) {
            c[idx] += val;
            idx += lowbit(idx);
        }
    }
    
    int sum(int idx) {
        idx = max(idx, 0);
        idx = min(idx, n);
        int s = 0;
        while (idx > 0) {
            s += c[idx];
            idx -= lowbit(idx);
        }
        return s;
    }
    
    ~BIT() {
        c.clear();
    }

private:
    int n;
    vector<int> c;
    
    int lowbit(int x) {
        return x & -x;
    }    
};

class Solution {
public:
    int countRangeSum(vector<int>& nums, int lower, int upper) {
        int n = nums.size();
        int64_t s = 0;
        map<int64_t, int> mm;
        mm[lower] = 0; mm[upper+1] = 0;
        
        int i, res = 0;
        for (i = 0; i < n; i++) {
            s += nums[i];
            mm[s] = 0;
            mm[s + lower] = 0;
            mm[s + upper + 1] = 0;
        }
        i = 0;
        for (auto it = mm.begin(); it != mm.end(); it++)
            it->second = ++i;
        
        BIT bit(i);
        
        bit.add(mm[lower], 1);
        bit.add(mm[upper+1], -1);
        s = 0;
        for (i = 0; i < n; i++) {
            s += nums[i];
            res += bit.sum(mm[s]);
            bit.add(mm[s + lower], 1);
            bit.add(mm[s + upper + 1], -1);
        }
        
        mm.clear();
        return res;
    }
};

File: python/327. Count of Range Sum.py
class Solution:
    def countRangeSum(self, nums: List[int], lower: int, upper: int) -> int:
        res, pre, now = 0, [0], 0
        for n in nums:
            now += n
            res += bisect.bisect_right(pre, now - lower) - bisect.bisect_left(pre, now - upper)
            bisect.insort(pre, now)
        return res


# 解题思路
# 使用前缀数组prepre，然后每个前缀和pre[i]pre[i]二分查找前面i-1i−1个和的pre[i]-lowerpre[i]−lower和pre[i]-upperpre[i]−upper的位置得出区间和的数量，然后把pre[i]pre[i]二分插入到数组中保持数组有序

# 作者：fan-cai
# 链接：https://leetcode-cn.com/problems/count-of-range-sum/solution/python3-6xing-dai-ma-jian-ji-qian-zhui-he-er-fen-c/
# 来源：力扣（LeetCode）
# 著作权归作者所有。商业转载请联系作者获得授权，非商业转载请注明出处。